### ISSUE TITLE
Specify epochs for hyperparameter optimization

### ISSUE THREAD

**Issue by peastman** [deepchem/deepchem#2194]
When you call `hyperparam_search()` on a `HyperparamOpt` object, it doesn't let you specify how many epochs of fitting to do on each model.  It just uses the default, which is 10 for KerasModel and TorchModel.  That isn't always an appropriate number.  We should add an option to specify the number of epochs to use.

**Comment by rbharath** [comment]
+1 to this! We should definitely allow control of number of epochs and other training hyperparameters

**Comment by cjgalvin** [comment]
This feature was also mentioned in this issue:
https://github.com/deepchem/deepchem/issues/2063

I used the modification suggested and it seems to work. It is weird to specify the nb_epochs as an optimization parameter, so adding it as an argument in `hyperparam_search` would be much clearer. The approach in the link may have been used due to a note above the `hyperparam_search` definition in `grid_search.py` indicating the number of arguments cannot be changed. I did not investigate, and hope this feature is implemented one or the other.

**Comment by rbharath** [comment]
@cjgalvin This is definitely on our radar! I've marked this issue as "help wanted" in case someone wants to take this on

**Comment by cjgalvin** [comment]
I wanted to follow up on how I implemented nb_epochs into the hyperparameter search function as a function argument, instead of as an optimization parameter. As I noted above, nb_epochs is not really an optimization parameter, so it is unnatural to assign its value in the optimization parameter dictionary. I only did this for the `GaussianProcessHyperparamOpt` class, but it should work for other optimization strategies in Deepchem assuming the API is consistent.

My modifications were:
1) Add nb_epochs as an argument to the `hyperparam_search` class function.
2) Defined a new class function called `_optimize`, which takes `nb_epochs` and '**placeholders` as arguments
3) Moved the contents of the `optimizing_function` function to `_optimize`, and rewrote `optimizing_function` to return `_optimize(nb_epochs, **placeholders)`

Now, the optimization function call looks something like this:
```
optimizer = dc.hyper.GaussianProcessHyperparamOpt(build_model_hp4dc)
opt_results = optimizer.hyperparam_search(
    params_dict=model_params, 
    nb_epochs=nb_epochs,
    train_dataset=train_dataset, 
    valid_dataset=val_dataset, 
    metric=metric, 
    logdir=log_dir
)
```
Where `nb_epochs` is defined by the user.

I am not sure if this solution is in keeping with Deepchem's intended use or style. If you provide feedback to that end, I can make modifications accordingly.

**Comment by rbharath** [comment]
@cjgalvin This looks to be a nice clean design fix! `nb_epochs` is indeed a little strange to specify as a hyperparameter to search over.

Would you consider submitting this as a PR? We would want to implement in for GridHyperparam search and in the abstract superclass API as well for consistency.

On a separate note, we might want to add some capability for selecting the best model checkpoint based on the validation set for hyperparameter turning. That often makes a difference in model validation performance. That's outside the scope of this change of course.

**Comment by cjgalvin** [comment]
@rbharath Re: PR, no problem. I'll  implement in GridHyperparam as well, then create the PR. 

I agree adding logic for selecting and doing some further action with the best performing model checkpoint would be useful, and is on my radar as well. If I implement anything, I'll create a new feature request with it.

**Comment by peastman** [comment]
Agreed, `nb_epochs` is a hyper-hyper-parameter, if you know what I mean.  It's a property of the hyperparameter optimization algorithm, not a property of the model it's searching over.

If you had plenty of time, the best thing to do would be to train each model until the validation set performance stopped improving.  But that could be really slow, so instead we settle for training each model for a fixed, not too large number of epochs.
